### PR TITLE
ARROW-17824: [C++][Gandiva] Implement preallocation for variable length output buffer

### PR DESCRIPTION
**Pull request by js8544**
When the output type of an expression is of variable length, e.g. string, Gandiva would realloc the output buffer to make space for new outputs for each row. When num of rows is high some memory allocators perform poorly.

We can use the std::vector like approach to amortize the allcation cost. First allocate some initial space depending on the input size. Each time we run out of space, double the buffer size. In the end shrink it to fit the actual size. Arrow string builder also uses this approach.

### CODE REVIEW

**Comment by js8544** [comment]
I've tested the performance with a newly added benchmark TimedTestOutputStringAllocs. 
On my machine, it changes from 361038us to 290553us for jemalloc, and 6765325us to 291024us. Huge improvement for mimalloc.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 1]
```diff
@@ -129,8 +129,22 @@ int32_t gdv_fn_populate_varlen_vector(int64_t context_ptr, int8_t* data_ptr,
...
 
-  // This also sets the size in the buffer.
-  auto status = buffer->Resize(offset + entry_len, false /*shrink*/);
+  auto new_size = offset + entry_len;
+  // preallocation, double the size to amortize costs
+  if (buffer->capacity() < new_size) {
+    auto status =
+        buffer->Reserve(std::max(buffer->capacity() * 2, static_cast<int64_t>(new_size)));
+    if (!status.ok()) {
+      gandiva::ExecutionContext* context =
```
Comment (kou) [inline cpp/src/gandiva/gdv_function_stubs.cc:138]: Can we use `auto` here?

File: cpp/src/gandiva/tests/micro_benchmarks.cc [hunk 2]
```diff
@@ -249,6 +250,31 @@ static void TimedTestAllocs(benchmark::State& state) {
...
+  auto length = TreeExprBuilder::MakeFunction("octet_length", {upper}, int32());
+  auto expr = TreeExprBuilder::MakeExpression(upper, field_res);
+
+  std::shared_ptr<Projector> projector;
+  ASSERT_OK(Projector::Make(schema, {expr}, TestConfiguration(), &projector));
+
+  FastUtf8DataGenerator data_generator(64);
+  ProjectEvaluator evaluator(projector);
+
+  Status status = TimedEvaluate<arrow::StringType, std::string>(
```
Comment (kou) [inline cpp/src/gandiva/tests/micro_benchmarks.cc:274]: Can we use `auto` here?

**Review by kou — APPROVED** [review]
+1

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 1 above)
Comment (js8544) [reply cpp/src/gandiva/gdv_function_stubs.cc]: done

File: cpp/src/gandiva/tests/micro_benchmarks.cc (hunk 2 above)
Comment (js8544) [reply cpp/src/gandiva/tests/micro_benchmarks.cc]: done

File: cpp/src/gandiva/tests/micro_benchmarks.cc [hunk 3]
```diff
@@ -249,6 +250,31 @@ static void TimedTestAllocs(benchmark::State& state) {
...
+
+  std::shared_ptr<Projector> projector;
+  ASSERT_OK(Projector::Make(schema, {expr}, TestConfiguration(), &projector));
+
+  FastUtf8DataGenerator data_generator(64);
+  ProjectEvaluator evaluator(projector);
+
+  auto status = TimedEvaluate<arrow::StringType, std::string>(
+      schema, evaluator, data_generator, pool_, 1 * MILLION, 16 * THOUSAND, state);
+  ASSERT_TRUE(status.ok());
```
Comment (kou) [inline cpp/src/gandiva/tests/micro_benchmarks.cc:276]: Ah, sorry. Can we use `ASSERT_OK()` here?

Comment (js8544) [reply cpp/src/gandiva/tests/micro_benchmarks.cc]: done

File: cpp/src/gandiva/tests/micro_benchmarks.cc [hunk 4]
```diff
@@ -249,6 +251,31 @@ static void TimedTestAllocs(benchmark::State& state) {
...
+
+  std::shared_ptr<Projector> projector;
+  ASSERT_OK(Projector::Make(schema, {expr}, TestConfiguration(), &projector));
+
+  FastUtf8DataGenerator data_generator(64);
+  ProjectEvaluator evaluator(projector);
+
+  auto status = TimedEvaluate<arrow::StringType, std::string>(
+      schema, evaluator, data_generator, pool_, 1 * MILLION, 16 * THOUSAND, state);
+  ASSERT_OK(status);
```
Comment (kou) [inline cpp/src/gandiva/tests/micro_benchmarks.cc:277]: Could you use the target expression directly with `ASSERT_OK()` to get better failure message?

```suggestion
  ASSERT_OK(TimedEvaluate<arrow::StringType, std::string>(
      schema, evaluator, data_generator, pool_, 1 * MILLION, 16 * THOUSAND, state));
```

Comment (js8544) [reply cpp/src/gandiva/tests/micro_benchmarks.cc:277]: I tried this but it doesn't compile.
```
/home/jinshang/arrow/cpp/src/gandiva/tests/micro_benchmarks.cc:312:138: error: macro "ASSERT_OK" passed 2 arguments, but takes just 1
  312 |   ASSERT_OK(TimedEvaluate<arrow::StringType, std::string>(schema, evaluator, data_generator, pool_, 100 * THOUSAND, 16 * THOUSAND, state));
      |                                                                                                                                          ^
In file included from /home/jinshang/arrow/cpp/src/gandiva/tests/micro_benchmarks.cc:21:
/home/jinshang/arrow/cpp/src/arrow/testing/gtest_util.h:87: note: macro "ASSERT_OK" defined here
   87 | #define ASSERT_OK(expr)                                                              \
      | 
/home/jinshang/arrow/cpp/src/gandiva/tests/micro_benchmarks.cc: In function ‘void gandiva::TimedTestMultiOr(benchmark::State&)’:
/home/jinshang/arrow/cpp/src/gandiva/tests/micro_benchmarks.cc:312:3: error: ‘ASSERT_OK’ was not declared in this scope
  312 |   ASSERT_OK(TimedEvaluate<arrow::StringType, std::string>(schema, evaluator, data_generator, pool_, 100 * THOUSAND, 16 * THOUSAND, state));
```
Seems to be linked to this: https://stackoverflow.com/questions/32241581/c-preprocessors-are-not-aware-of-template-arguments
So I wrapped an additional pair of parentheses and it works now.

**Comment by kou** [comment]
Thanks!

**Pull request merged**